### ISSUE TITLE
The performance is very BAD, please improve it

### ISSUE THREAD

**Issue by yourchanges** [rana/ora#86]
Hi,

I have write a simple code to fetch the **2000000** rows from table use three golang oracle driver:

- gopkg.in/goracle.v1                  **==> about 22 seconds**
- github.com/mattn/go-oci8         **==> about 34 seconds**
- gopkg.in/rana/ora.v3                **==> about 660 seconds**

**My ENV:**
centos 6.6 64bit
golang 1.5.3

**The code:**

`func FetchData(db *sql.DB) []int {

	var r []int

	sqlstr := `SELECT id FROM xxxx.geolocation`

	// Prepare statement for reading data
	stmtOut, err := db.Prepare(sqlstr)
	checkErr(err)
	defer stmtOut.Close()

	rows, err := stmtOut.Query()
	checkErr(err)

	for rows.Next() {
		var id int
		err = rows.Scan(&id)
		checkErr(err)
		r = append(r, id)
	}
	return r
}`

**Comment by tgulacsi** [comment]
What kind of table is that? Can you share its CREATE TABLE statement?
Maybe a small (10 record) sample, too?
What is the version of the library? 8ad7aa7c4a597f53cd08e5d267ac3af0d74ad379 is needed for prefetching to work, which is crucial here.

**Comment by yourchanges** [comment]
CREATE TABLE "xxxx"."GEOLOCATION" 
   (	"ID" NUMBER(*,0) NOT NULL ENABLE, 
	"PERSON_ID" NUMBER(*,0), 
	"PERSON_ACCOUNT_ID" NUMBER(*,0), 
	"ORGANIZATION_ID" NUMBER(*,0), 
	"ORGANIZATION_MEMBERSHIP_ID" NVARCHAR2(45), 
	"LOCATION" NVARCHAR2(2000) NOT NULL ENABLE, 
	"DEVICE_ID" NVARCHAR2(45), 
	"DEVICE_REGISTRATION_ID" NVARCHAR2(500), 
	"DEVICE_NAME" NVARCHAR2(45), 
	"DEVICE_TYPE" NVARCHAR2(45), 
	"DEVICE_OS_NAME" NVARCHAR2(45), 
	"DEVICE_TOKEN" NVARCHAR2(45), 
	"DEVICE_OTHER_DETAILS" NVARCHAR2(100), 
	"ADDED_TIME" DATE NOT NULL ENABLE, 
	"ROW_CHANGED_TIME" TIMESTAMP (6) NOT NULL ENABLE, 
	 CONSTRAINT "GEOLOCATION_PK" PRIMARY KEY ("ID")
   )

   "ID","PERSON_ID","PERSON_ACCOUNT_ID","ORGANIZATION_ID","ORGANIZATION_MEMBERSHIP_ID","LOCATION","DEVICE_ID","DEVICE_REGISTRATION_ID","DEVICE_NAME","DEVICE_TYPE","DEVICE_OS_NAME","DEVICE_TOKEN","DEVICE_OTHER_DETAILS","ADDED_TIME","ROW_CHANGED_TIME","RN"
"1","8.37064876162908E16","8.37064898728264E16","12","6506","POINT(30.5518407 104.0685472)","a71223186cef459b","","Samsung SCH-I545","Mobile","Android 4.4.2","","","2015-09-14 23:42:24","27-APR-16 11.09.53.000000 PM","1"
"2","8.37064876162908E16","8.37064898728264E16","12","6506","POINT(30.5520498 104.0686355)","a71223186cef459b","","Samsung SCH-I545","Mobile","Android 4.4.2","","","2015-09-14 23:44:31","27-APR-16 11.09.53.000000 PM","2"
"3","8.37064876162908E16","8.37064898728264E16","12","6506","POINT(30.5517747 104.0684895)","a71223186cef459b","","Samsung SCH-I545","Mobile","Android 4.4.2","","","2015-09-14 23:45:23","27-APR-16 11.09.53.000000 PM","3"
"4","8.64522675633357E16","8.64522734353613E16","","1220457","POINT(30.55187 104.06856)","3A9D1838-3B2D-4119-9E07-77C6CDAC53C5","noUwBnWojdY:APA91bE8aGLEECS9_Q1EKrp8i2B36H1X8GwIj3v58KUcuXglhf0rXJb8Ez5meQ6D5MgTAQghYEe3s9vOntU3pYPQoc6ASNw3QzhzQevAqlMQC2ukUMNyLD8Rve-IA1-6lttsCXYsYIKh","User3’s iPhone","iPhone","iPhone OS","","DeviceID:3A9D1838-3B2D-4119-9E07-77C6CDAC53C5, SystemVersion:8.4, LocalizedModel:iPhone","2015-09-14 23:45:29","27-APR-16 11.09.53.000000 PM","4"
"5","8.37064876162908E16","8.37064898728264E16","12","6506","POINT(30.5517458 104.0685809)","a71223186cef459b","","Samsung SCH-I545","Mobile","Android 4.4.2","","","2015-09-14 23:46:23","27-APR-16 11.09.53.000000 PM","5"
"6","8.37064876162908E16","8.37064898728264E16","12","6506","POINT(30.551802 104.0685301)","a71223186cef459b","","Samsung SCH-I545","Mobile","Android 4.4.2","","","2015-09-14 23:47:23","27-APR-16 11.09.53.000000 PM","6"
"7","8.64522675633357E16","8.64522734353613E16","","1220457","POINT(30.55187 104.06856)","3A9D1838-3B2D-4119-9E07-77C6CDAC53C5","noUwBnWojdY:APA91bE8aGLEECS9_Q1EKrp8i2B36H1X8GwIj3v58KUcuXglhf0rXJb8Ez5meQ6D5MgTAQghYEe3s9vOntU3pYPQoc6ASNw3QzhzQevAqlMQC2ukUMNyLD8Rve-IA1-6lttsCXYsYIKh","User3’s iPhone","iPhone","iPhone OS","","DeviceID:3A9D1838-3B2D-4119-9E07-77C6CDAC53C5, SystemVersion:8.4, LocalizedModel:iPhone","2015-09-14 23:47:29","27-APR-16 11.09.53.000000 PM","7"
"8","8.37064876162908E16","8.37064898728264E16","12","6506","POINT(30.551952 104.0685893)","a71223186cef459b","","Samsung SCH-I545","Mobile","Android 4.4.2","","","2015-09-14 23:49:18","27-APR-16 11.09.53.000000 PM","8"
"9","8.37064876162908E16","8.37064898728264E16","12","6506","POINT(30.5518439 104.0685473)","a71223186cef459b","","Samsung SCH-I545","Mobile","Android 4.4.2","","","2015-09-14 23:49:48","27-APR-16 11.09.53.000000 PM","9"
"10","8.37064876162908E16","8.37064898728264E16","12","6506","POINT(30.5518439 104.0685473)","a71223186cef459b","","Samsung SCH-I545","Mobile","Android 4.4.2","","","2015-09-14 23:49:51","27-APR-16 11.09.53.000000 PM","10"

**Comment by yourchanges** [comment]
your_@leewin10 MINGW64 /d/goworkspace/src/gopkg.in/rana/ora.v3 (master)
$ git log -1
commit 3b4c86df1dd1e21dbaeaa4ea9827e37f12903702
Author: Tamás Gulácsi <tgulacsi78@gmail.com>
Date:   Wed Mar 9 10:55:09 2016 +0100

    don't panic on nil with logging

    sysName should not panic when anything is nil.

    Fixes #68.

**Comment by yourchanges** [comment]
After upgrade to 

`git log -1
commit 68741860f1e451d10fb5a9cc211c7b85285a8300
Author: Tamás Gulácsi <tgulacsi78@gmail.com>
Date:   Thu Apr 7 20:08:18 2016 +0200

    go vet: defRset fix of ocistmt
`

the execute time is about **34 seconds**, but the python cx_Oracle only takes 18 seconds to do the same thing

**Comment by tgulacsi** [comment]
So, it is on par with go-oci8? That's great!

I'll create a benchmark, so we can inspect this in detail. My gut feeling says that this has to be the NUMBER(,0) column - though now that's fetched from Oracle as OCINum (no conversion), the conversion to int (id) is not optimized: it is converted to String, then int.

defOCINum.value() returns the string, as that's the most precise thing to do (neither an int64, nor a float64 can hold the OCINum's 38 decimal digits). I don't know yet how to force it to return an int64 or float64 if it fits, or if that's desired.

But first we should measure - maybe not even this is the bottleneck!

**Comment by tgulacsi** [comment]
If you'd help me to profile the BenchmarkSelect benchmark on the "issue86" branch, that'd be great!
I'll be back on this on Tuesday.

**Comment by tgulacsi** [comment]
`go test -c && ./ora.v3.test -test.run=^$ -test.bench=Select -test.cpuprofile /tmp/cpu.prof && go tool pprof -cum -top ./ora.v3.test /tmp/cpu.prof

With 7d512f6, benchmark is down from 257210201ns to 186406940ns.
But I don't understand why asString() is there at all.

**Comment by tgulacsi** [comment]
As I see now with `go tool pprof -cum -lines -png ./ora.v3.test /tmp/cpu.prof  >/tmp/x.png` 
![x](https://cloud.githubusercontent.com/assets/256074/15303477/4d7a9f6e-1bb8-11e6-9f60-1e0272298ee1.png)

The most time consuming is calling OCIStmtFetch2. http://www.sqlines.com/oracle/oci/array_fetch says that we could instrument that call to return whole arrays, not fields one-by-one, but that needs a little bit more work.
As cgo calls are not free, and a SELECT with a few thousand records does mostly that, if performance is a goal, we should make that step.

**Comment by rana** [comment]
I wonder if prefetch tuning affects the time for this

**Comment by tgulacsi** [comment]
We already have prefetch enabled, so the next candidate is less C calls, in my opinion.

**Comment by vmogilev** [comment]
We can tell OCIStmtFetch2 to fetch more than 1 row at a time.  

It looks like currently the library does row by row fetching (which is slow): https://github.com/rana/ora/blob/71162529d1973f392a89abdd851bdc7f1de75db2/rset.go#L168

Here's a C example:

```
/*
NAME
ocidemo.c - Demonstrates connecting through a TNS name (possibly to
an Oracle Rdb database) and executing SQL statements.
*/

/* This program accepts arbitrary SQL statements from the user,
and processes the statement. Statements may be entered on
multiple lines, and must be terminated by a semi-colon.
If a query, the results are printed.

Statements are entered at the OCISQL prompt.

To quit the program, type EXIT at the OCISQL prompt.
*/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <oci.h>

/* Constants used in this program. */
#define MAX_BINDS 12
#define MAX_ITEM_BUFFER_SIZE 33
#define MAX_SELECT_LIST_SIZE 12
#define MAX_SQL_IDENTIFIER 31
#define MAX_FETCH_COUNT 15

/* Declare this programs functions. */
sword connect_user();
sword get_sql_statement();
static void checkerr(/*_ OCIError *errhp, sword status _*/);
void do_exit();

sword describe_define();
sword do_binds();
void print_header();
void print_rows();

/* Globals */
static OCIEnv* envhp;
static OCIError* errhp;
static OCISvcCtx* svchp;
static OCIStmt* stmthp;

static text sql_statement[2048];
static ub2 sql_function;
static sword numwidth = 8;

/* Declare an array of bind values. */
text bind_values[MAX_BINDS][MAX_ITEM_BUFFER_SIZE];
/* Declare an array of pointers to bind handles. */
OCIBind* bindhp[MAX_BINDS];
sb2 bind_ind = 0;

/* Declare structures for query information. */
struct describe {
  sb4 dbsize;
  sb2 dbtype;
  OraText buf[MAX_ITEM_BUFFER_SIZE];
  ub4 buflen;
  sb1 scale;
};

struct define {
  OraText buf[MAX_FETCH_COUNT][MAX_ITEM_BUFFER_SIZE];
  float flt_buf[MAX_FETCH_COUNT];
  sb4 int_buf[MAX_FETCH_COUNT];
  sb2 indp[MAX_FETCH_COUNT];
  OCIDefine* defhp;
};

/* Define arrays of describe and define structs. */
struct describe desc[MAX_SELECT_LIST_SIZE];
struct define def[MAX_SELECT_LIST_SIZE];

int main(argc, argv) int argc;
char* argv[];
{
  sword ncols;
  sword errcode = 0;
  ub2 rows_processed = 0;
  ub4 prefetch_cnt;

  /* Create and initialize environment handle */
  errcode = OCIEnvCreate((OCIEnv**)&envhp, (ub4)OCI_DEFAULT, (dvoid*)0,
                         (dvoid * (*)(dvoid*, size_t))0,
                         (dvoid * (*)(dvoid*, dvoid*, size_t))0,
                         (void (*)(dvoid*, dvoid*))0, (size_t)0, (dvoid**)0);

  if (errcode != 0) {
    (void)printf("OCIEnvCreate failed with errcode = %d.\n", errcode);
    exit(1);
  }

  /* Allocate error handle */
  (void)OCIHandleAlloc((dvoid*)envhp, (dvoid**)&errhp, OCI_HTYPE_ERROR,
                       (size_t)0, (dvoid**)0);

  /* Allocate service context */
  (void)OCIHandleAlloc((dvoid*)envhp, (dvoid**)&svchp, OCI_HTYPE_SVCCTX,
                       (size_t)0, (dvoid**)0);

  /* Connect */
  if (connect_user()) exit(-1);

  /* Allocate statment handle */
  checkerr(errhp, OCIHandleAlloc((dvoid*)envhp, (dvoid**)&stmthp,
                                 OCI_HTYPE_STMT, (size_t)0, (dvoid**)0));

  /* Process user's SQL statements. */
  for (;;) {
    /* Get the statement, exit on "exit". */
    if (get_sql_statement()) {
      printf("\n");
      do_exit(0);
    }

    /* Prepare the statement. This is a purely local call. */
    if (errcode = OCIStmtPrepare(stmthp, errhp, sql_statement,
                                 (ub4)strlen((char*)sql_statement),
                                 (ub4)OCI_NTV_SYNTAX, (ub4)OCI_DEFAULT)) {
      printf("\nError from OCIStmtPrepare\n");
      checkerr(errhp, errcode);
      continue;
    }

    /* Get the statment type */
    checkerr(errhp, OCIAttrGet((dvoid*)stmthp, (ub4)OCI_HTYPE_STMT,
                               (dvoid*)&sql_function, (ub4*)0,
                               (ub4)OCI_ATTR_STMT_TYPE, (OCIError*)errhp));

    /* Bind any input variables. */
    if ((ncols = do_binds(sql_statement)) == -1) continue;

    /* If the statement is a query, set the prefetch then describe and define
all select-list items before doing the execute. */
    if (sql_function == OCI_STMT_SELECT) {
      /* Set prefetch to 20 rows */
      prefetch_cnt = 20;
      OCIAttrSet((dvoid*)stmthp, (ub4)OCI_HTYPE_STMT, (dvoid*)&prefetch_cnt,
                 (ub4)0, (ub4)OCI_ATTR_PREFETCH_ROWS, (OCIError*)errhp);

      if ((ncols = describe_define()) == -1) continue;
    }

    /* Execute the statement. */
    if (errcode = OCIStmtExecute(svchp, stmthp, errhp, 1, 0, (OCISnapshot*)0,
                                 (OCISnapshot*)0, OCI_DEFAULT)) {
      printf("\nError from OCIStmtExecute\n");
      checkerr(errhp, errcode);
      continue;
    }

    /* Fetch and display the rows for the query. */
    if (sql_function == OCI_STMT_SELECT) {
      print_header(ncols);
      print_rows(ncols);
    }

    /* Print the rows-processed count. */
    if (sql_function == OCI_STMT_SELECT || sql_function == OCI_STMT_UPDATE ||
        sql_function == OCI_STMT_DELETE || sql_function == OCI_STMT_INSERT) {
      /* Get the number of rows processed */
      checkerr(errhp, OCIAttrGet((dvoid*)stmthp, (ub4)OCI_HTYPE_STMT,
                                 (dvoid*)&rows_processed, (ub4*)0,
                                 OCI_ATTR_ROW_COUNT, (OCIError*)errhp));

      printf("\n%d row%c processed.\n", rows_processed,
             rows_processed == 1 ? ' ' : 's');
    } else
      printf("\nStatement processed.\n");

  } /* end for (;;) */

} /* end main() */

sword connect_user() {
  /* Only using a single user session so can use OCILogon */

  text username[132];
  text password[132];
  text tns_name[132];
  sword n;

  /* Three tries to connect. */
  for (n = 3; --n >= 0;) {
    printf("TNS name: ");
    gets((char*)tns_name);
    printf("Username: ");
    gets((char*)username);
    printf("Password: ");
    gets((char*)password);

    if (OCILogon(envhp, errhp, &svchp, username, (ub4)strlen((char*)username),
                 password, (ub4)strlen((char*)password), tns_name,
                 (ub4)strlen((char*)tns_name))) {
      printf("Cannot connect to %s as %s.\n", tns_name, username);
      printf("Try again.\n\n");
    } else {
      return 0;
    }
  }
  printf("Connection failed. Exiting...\n");
  return -1;
}

/* Describe select-list items. */

sword describe_define() {
  ub4 col, deflen, deftyp;
  static ub1* defptr;
  OCIParam* colhd = (OCIParam*)0; /* column handle */
  OraText* col_name;
  ub4 numcols = 0;
  sword errcode = 0;

  /* Describe the statement */
  if (errcode = OCIStmtExecute(svchp, stmthp, errhp, 0, 0, (OCISnapshot*)0,
                               (OCISnapshot*)0, OCI_DESCRIBE_ONLY)) {
    checkerr(errhp, errcode);
    return (-1);
  }

  /* Get the number of columns in the query */
  checkerr(errhp, OCIAttrGet((dvoid*)stmthp, OCI_HTYPE_STMT, (dvoid*)&numcols,
                             (ub4*)0, OCI_ATTR_PARAM_COUNT, errhp));

  if (numcols > MAX_SELECT_LIST_SIZE) {
    printf("Too many select list items - %d", numcols);
    return (-1);
  }

  /* Get descriptions of the select-list items. */
  for (col = 0; col < numcols; col++) {
    /* Get parameter for column, column positions start from 1 */
    checkerr(errhp, OCIParamGet((dvoid*)stmthp, OCI_HTYPE_STMT, errhp,
                                (dvoid**)&colhd, (col + 1)));

    /* Get data-type of column */
    checkerr(errhp, OCIAttrGet((dvoid*)colhd, OCI_DTYPE_PARAM,
                               (dvoid*)&desc[col].dbtype, (ub4*)0,
                               OCI_ATTR_DATA_TYPE, errhp));
    /* Get other attributes */
    checkerr(errhp, OCIAttrGet((dvoid*)colhd, OCI_DTYPE_PARAM,
                               (dvoid*)&desc[col].dbsize, (ub4*)0,
                               OCI_ATTR_DATA_SIZE, errhp));

    checkerr(errhp, OCIAttrGet((dvoid*)colhd, OCI_DTYPE_PARAM,
                               (dvoid*)&desc[col].scale, (ub4*)0,
                               OCI_ATTR_SCALE, errhp));

    col_name = 0;
    checkerr(errhp, OCIAttrGet((dvoid*)colhd, OCI_DTYPE_PARAM, &col_name,
                               &desc[col].buflen, OCI_ATTR_NAME, errhp));
    strncpy((char*)desc[col].buf, (char*)col_name, MAX_ITEM_BUFFER_SIZE);

    /* Initialize define handle */
    def[col].defhp = (OCIDefine*)0;

    /* adjust sizes and types for display */
    switch (desc[col].dbtype) {
      case SQLT_NUM:
        desc[col].dbsize = numwidth;
        /* Handle NUMBER with scale as float. */
        if (desc[col].scale != 0) {
          defptr = (ub1*)&def[col].flt_buf[0];
          deflen = (sword)sizeof(float);
          deftyp = SQLT_FLT;
          desc[col].dbtype = SQLT_FLT;
        } else {
          defptr = (ub1*)&def[col].int_buf[0];
          deflen = sizeof(sb4);
          deftyp = SQLT_INT;
          desc[col].dbtype = SQLT_INT;
        }
        break;
      default:
        if (desc[col].dbtype == SQLT_DATE) desc[col].dbsize = 9;
        if (desc[col].dbtype == SQLT_RDD) /* ROWID */
          desc[col].dbsize = 18;
        defptr = def[col].buf[0];
        deflen = MAX_ITEM_BUFFER_SIZE;
        deftyp = SQLT_STR; /* NULL-terminated STRING */
        break;
    }

    /* Tell OCI where to put data that is fetched for this column */
    if (errcode = OCIDefineByPos(stmthp, &def[col].defhp, errhp, (ub4)(col + 1),
                                 defptr, deflen, deftyp, &def[col].indp[0],
                                 (ub2*)0, (ub2*)0, OCI_DEFAULT)) {
      checkerr(errhp, errcode);
      printf("Error on define of col %d\n", (col + 1));
      printf("dbtype=%d, deftyp=%d, deflen=%d\n", desc[col].dbtype, deftyp,
             deflen);
      return -1;
    }
  }
  return col;
}

/* Bind input variables. */

sword do_binds(stmt_buf) text* stmt_buf;
{
  sword i, in_literal, n, errcode;
  text *cp, *ph;

  /* Find and bind input variables for placeholders. */
  for (i = 0, in_literal = FALSE, cp = stmt_buf; *cp && i < MAX_BINDS; cp++) {
    if (*cp == '\'') in_literal = ~in_literal;
    /* Look for input variables beginning with ':' */
    if (*cp == ':' && !in_literal) {
      for (ph = ++cp, n = 0;
           *cp && (isalnum(*cp) || *cp == '_') && n < MAX_SQL_IDENTIFIER;
           cp++, n++)
        ;
      *cp = '\0';
      printf("Enter value for %s: ", ph);
      gets((char*)&bind_values[i][0]);

      /* Bind using the placeholder name ph
NOTE: the bind variable address must be static.
This would not work if bind_values were an
auto on the do_binds stack. */
      if (errcode = OCIBindByName(stmthp, &bindhp[i], errhp, ph,
                                  strlen((char*)ph), (ub1*)&bind_values[i][0],
                                  (strlen((char*)bind_values[i]) + 1), SQLT_STR,
                                  &bind_ind, (ub2*)0, (ub2)0, (ub4)0, (ub4*)0,
                                  OCI_DEFAULT)) {
        checkerr(errhp, errcode);
        return -1;
      }
      i++;
    } /* end if (*cp == ...) */
  }   /* end for () */
  return i;
}

/* Clean up and exit. */
void do_exit(rv) sword rv;
{
  checkerr(errhp, OCILogoff((dvoid*)svchp, errhp));
  exit(rv);
}

sword get_sql_statement() {
  text cbuf[1024];
  text* cp;
  sword stmt_level;

  for (stmt_level = 1;;) {
    if (stmt_level == 1) {
      /* Init statement buffer and print prompt. */
      *sql_statement = '\0';
      fputs("\nOCISQL> ", stdout);
    } else {
      printf("%3d ", stmt_level);
    }

    /* Get (part of) a SQL statement. */
    gets((char*)cbuf);
    if (*cbuf == '\0') continue;
    if (strncmp((char*)cbuf, "exit", 4) == 0) return -1;

    /* Concatenate to statement buffer. */
    if (stmt_level > 1) strcat((char*)sql_statement, " ");
    strcat((char*)sql_statement, (char*)cbuf);

    /* Check for possible terminator. */
    cp = &sql_statement[strlen((char*)sql_statement) - 1];

    while (isspace(*cp)) cp--;
    if (*cp == ';') {
      *cp = '\0';
      break;
    }
    stmt_level++;
  }
  return 0;
}

void print_header(ncols) sword ncols;
{
  sword col, n;

  fputc('\n', stdout);

  for (col = 0; col < ncols; col++) {
    n = desc[col].dbsize - desc[col].buflen;
    if (desc[col].dbtype == SQLT_FLT || desc[col].dbtype == SQLT_INT) {
      printf("%*c", n, ' ');
      printf("%*.*s", desc[col].buflen, desc[col].buflen, desc[col].buf);
    } else {
      printf("%*.*s", desc[col].buflen, desc[col].buflen, desc[col].buf);
      printf("%*c", n, ' ');
    }
    fputc(' ', stdout);
  }
  fputc('\n', stdout);

  for (col = 0; col < ncols; col++) {
    for (n = desc[col].dbsize; --n >= 0;) fputc('-', stdout);
    fputc(' ', stdout);
  }
  fputc('\n', stdout);
}

void print_rows(ncols) sword ncols;
{
  /* OCIStmtExecute will normally have fetched the first row */

  sword col, n, m;
  sword status = OCI_SUCCESS;
  ub4 rows_fetched;
  m = 0;

  do {
    /* Get the number of rows returned by the last fetch or execute */
    checkerr(errhp, OCIAttrGet((dvoid*)stmthp, (ub4)OCI_HTYPE_STMT,
                               (dvoid*)&rows_fetched, (ub4*)0,
                               OCI_ATTR_ROWS_FETCHED, (OCIError*)errhp));
    if (rows_fetched > MAX_FETCH_COUNT) rows_fetched = MAX_FETCH_COUNT;

    for (m = 0; m < rows_fetched; m++) {
      fputc('\n', stdout);
      for (col = 0; col < ncols; col++) {
        /* Check col. return code for null. If
null, print n spaces, else print value. */
        if (def[col].indp[m] < 0)
          printf("%*c", desc[col].dbsize, '?');
        else {
          switch (desc[col].dbtype) {
            case SQLT_FLT:
              printf("%*.*f", numwidth, 2, def[col].flt_buf[m]);
              break;
            case SQLT_INT:
              printf("%*d", numwidth, def[col].int_buf[m]);
              break;
            default:
              printf("%s", def[col].buf[m]);
              n = desc[col].dbsize - strlen((char*)def[col].buf[m]);
              if (n > 0) printf("%*c", n, ' ');
              break;
          }
        }
        fputc(' ', stdout);
      }
    }
    /* With pre-fetch can get some rows even with OCI_NO_DATA status
so another pass is done to print those rows.
Break before doing another fetch. */
    if (status == OCI_NO_DATA) break;
  } while ((status = OCIStmtFetch2(stmthp, errhp, MAX_FETCH_COUNT,
                                   OCI_FETCH_NEXT, 1, OCI_DEFAULT)) ==
               OCI_SUCCESS ||
           status == OCI_SUCCESS_WITH_INFO || status == OCI_NO_DATA);

  if (status == OCI_ERROR) checkerr(errhp, status);
}

void checkerr(errhp, status) OCIError* errhp;
sword status;
{
  text errbuf[512];
  sb4 errcode = 0;

  switch (status) {
    case OCI_SUCCESS:
      break;
    case OCI_SUCCESS_WITH_INFO:
      (void)printf("Error - OCI_SUCCESS_WITH_INFO\n");
      break;
    case OCI_NEED_DATA:
      (void)printf("Error - OCI_NEED_DATA\n");
      break;
    case OCI_NO_DATA:
      (void)printf("Error - OCI_NODATA\n");
      break;
    case OCI_ERROR:
      (void)OCIErrorGet((dvoid*)errhp, (ub4)1, (text*)NULL, &errcode, errbuf,
                        (ub4)sizeof(errbuf), OCI_HTYPE_ERROR);
      (void)printf("Error - %.*s\n", 512, errbuf);
      break;
    case OCI_INVALID_HANDLE:
      (void)printf("Error - OCI_INVALID_HANDLE\n");
      break;
    case OCI_STILL_EXECUTING:
      (void)printf("Error - OCI_STILL_EXECUTE\n");
      break;
    case OCI_CONTINUE:
      (void)printf("Error - OCI_CONTINUE\n");
      break;
    default:
      break;
  }
}

```
**REF: ML Doc ID 1126015.1**

**Comment by tgulacsi** [comment]
A very rough proof of concept (ONLY with OCINum and Int64 columns, nothing tested, only the Select benchmark):

Old:
BenchmarkSelect-4         100000             23458 ns/op

New, with length 100 array:
BenchmarkSelect-4         100000             18599 ns/op

With 1000 length array:
BenchmarkSelect-4         100000             12432 ns/op

But these results are very flaky.

@yourchanges can you test it to have some comparison numbers against your original select?

**Comment by tgulacsi** [comment]
By the way, I'm tired of manually altering all def[UI]nt{8,16,32,64}.go and defFloat{32,64}.go, so I've implemeneted a very simple "generator" (sed-like) using defInt64.go and defFloat64.go as templates.
This way they will always be the same, and easier to maintain.

This could be expanded to generate bnd[UI]nt{8,16,32,64}{,Ptr,Slice}.go from bndInt64{,Ptr,Slice}.go, but first I want to test this approach.

**Comment by rana** [comment]
Nice

**Comment by yourchanges** [comment]
Very Very impressed & cool !!!

I have do the same test based on branch issue86, I got:

**Fetching 2000000 rows in 6 seconds.
Fetching 5000000 rows in 15 seconds.**

and after several repeat test, sometime the result gets "higher crest elevation", but I can't tell that it cased by the network or oracle DB high load.

**Comment by tgulacsi** [comment]
Hi all!

I'd like to ask for a helping hand: the "issue86" branch miserably fails if fetchArrLen>1 in rset.go.
Can someone help, what's the problem?

**Comment by tgulacsi** [comment]
Ok, now the issue86 branch passes the tests IFF fetchArrLen=1.
Still fails somewhere in rset.Next if fetchArrLen>1.

**Comment by tgulacsi** [comment]
Hi all!

Finally I've been able to fix my own errors, which stems from my lack of C knowledge. For the future myself: sizeof_dvoid is NOT the size of the void pointer!

So now the "issue86" branch passes all the tests that "master" passes!
It has two constant knobs: MaxFetchLen and MinFetchLen. Iff the query results in a LOB/LONG column, then MinFetchLen will be used (that much memory will be allocated) - otherwise MaxFetchLen is used. This is to be able to pass the tests - with 128 LOBs I couldn't run it.

TL;DR; *please* test this branch, as I want to merge it after all!

(Two flukes: TestStmt_Exe_select and TestSession_Tx_StartCommit fails when I run all tests, but passes on their own...)
